File: python/delayed-pi-exchange/pisender.py
# -*- coding: utf-8 -*-
import time
import helics as h
from math import pi

initstring = "2 --name=mainbroker"
fedinitstring = "--broker=mainbroker --federates=1"
timePeriod = 1

helicsversion = h.helicsGetVersion()

print("PI SENDER: Helics version = {}".format(helicsversion))

# Create broker #
print("Creating Broker")
broker = h.helicsCreateBroker("zmq", "", initstring)
print("Created Broker")

print("Checking if Broker is connected")
isconnected = h.helicsBrokerIsConnected(broker)
print("Checked if Broker is connected")

if isconnected == 1:
    print("Broker created and connected")

# Create Federate Info object that describes the federate properties #
fedinfo = h.helicsFederateInfoCreate()

# Set Federate name #
status = h.helicsFederateInfoSetFederateName(fedinfo, "TestA Federate")

# Set core type from string #
status = h.helicsFederateInfoSetCoreTypeFromString(fedinfo, "zmq")

# Federate init string #
status = h.helicsFederateInfoSetCoreInitString(fedinfo, fedinitstring)

# Set the message interval (timedelta) for federate. Note th#
# HELICS minimum message time interval is 1 ns and by default
# it uses a time delta of 1 second. What is provided to the
# setTimedelta routine is a multiplier for the default timedelta.

# Set one second message interval #
status = h.helicsFederateInfoSetPeriod(fedinfo, timePeriod)

status = h.helicsFederateInfoSetLoggingLevel(fedinfo, 1)

# Create value federate #
fed = h.helicsCreateCombinationFederate(fedinfo)
print("PI SENDER: Value federate created")

# Register the publication #
pub = h.helicsFederateRegisterGlobalPublication(fed, "testA", "double", "")
print("PI SENDER: Publication registered")

epid = h.helicsFederateRegisterGlobalEndpoint(fed, "endpoint1", "")

fid = h.helicsFederateRegisterSourceFilter(
    fed, h.helics_delay_filter, "endpoint1", "filter-name"
)

h.helicsFilterSet(fid, "delay", 2.0)

# Enter execution mode #
status = h.helicsFederateEnterExecutionMode(fed)
print("PI SENDER: Entering execution mode")

# This federate will be publishing deltat*pi for numsteps steps #
this_time = 0.0
value = pi

for t in range(5, 10):
    val = value

    currenttime = h.helicsFederateRequestTime(fed, t)

    status = h.helicsPublicationPublishDouble(pub, val)
    print(
        "PI SENDER: Sending value pi = {} at time {} to PI RECEIVER".format(
            val, currenttime[-1]
        )
    )

    status = h.helicsEndpointSendMessageRaw(epid, "endpoint2", str(t))

    time.sleep(1)

status = h.helicsFederateFinalize(fed)
print("PI SENDER: Federate finalized")

while h.helicsBrokerIsConnected(broker):
    time.sleep(1)

h.helicsFederateFree(fed)
h.helicsCloseLibrary()

print("PI SENDER: Broker disconnected")

File: c/pi-exchange/pi_sender.c
/*
Copyright © 2017-2019,
Battelle Memorial Institute; Lawrence Livermore National Security, LLC; Alliance for Sustainable Energy, LLC.  See
the top-level NOTICE for additional details. All rights reserved.
SPDX-License-Identifier: BSD-3-Clause
*/
static char help[] = " PI SENDER: Simple program to demonstrate the usage of HELICS C Interface.\n\
            This example creates a ZMQ broker and a value federate.\n\
            The value federate creates a global publications and publishes\n\
            t*pi for 20 time-steps with a time-step of 0.01 seconds.\n\n";


#include <stdio.h>
#include <helics/shared_api_library/ValueFederate.h>
#ifdef _MSC_VER
#include <windows.h>
#else
#include <unistd.h>
#endif

int main()
{
  helics_federate_info fedinfo;
  const char*    helicsversion;
  /*helics_status   status; */
  helics_broker  broker;
  const char*    initstring="-f 2 --name=mainbroker";
  const char*    fedinitstring="--broker=mainbroker --federates=1";
  int            isconnected;
  double         deltat=0.01;
  helics_federate vfed;
  helics_publication pub;
  double value = 22.0 / 7.0;
  helics_time currenttime = 0.0;
  int           numsteps = 20, i;
  helics_error err = helicsErrorInitialize();

  helicsversion = helicsGetVersion();

  printf("PI SENDER: Helics version = %s\n",helicsversion);
  printf("%s",help);

  /* Create broker */
  broker = helicsCreateBroker("zmq","",initstring,NULL);

  isconnected = helicsBrokerIsConnected(broker);

  if(isconnected) {
    printf("PI SENDER: Broker created and connected\n");
  }

  /* Create Federate Info object that describes the federate properties */
  fedinfo = helicsCreateFederateInfo();

  /* Set core type from string */
   helicsFederateInfoSetCoreTypeFromString(fedinfo,"zmq",NULL);

  /* Federate init string */
  helicsFederateInfoSetCoreInitString(fedinfo,fedinitstring,NULL);

  /* Set the message interval (timedelta) for federate. Note that
     HELICS minimum message time interval is 1 ns and by default
     it uses a time delta of 1 second. What is provided to the
     setTimedelta routine is a multiplier for the default timedelta.
  */
  /* Set one second message interval */
  helicsFederateInfoSetTimeProperty(fedinfo, helics_property_time_period, deltat, &err);
  helicsFederateInfoSetIntegerProperty(fedinfo, helics_property_int_log_level, 1, &err);


  /* Create value federate */
  vfed = helicsCreateValueFederate("Test sender Federate",fedinfo,NULL);
  printf("PI SENDER: Value federate created\n");

  /* Register the publication */
  pub = helicsFederateRegisterGlobalPublication(vfed,"testA",helics_data_type_double,"",NULL);
  printf("PI SENDER: Publication registered\n");

  /* Enter initialization mode */
   helicsFederateEnterInitializingMode(vfed,NULL);
  printf("PI SENDER: Entered initialization mode\n");

  /* Enter execution mode */
   helicsFederateEnterExecutingMode(vfed,NULL);
  printf("PI SENDER: Entered execution mode\n");

  /* This federate will be publishing deltat*pi for numsteps steps */


  for(i=0; i < numsteps; i++) {
    double val = currenttime*value;

    printf("PI SENDER: Sending value %3.2fpi = %4.3f at time %3.2f to PI RECEIVER\n",deltat*i,val,currenttime);
    helicsPublicationPublishDouble(pub,val,NULL);

    currenttime=helicsFederateRequestTime(vfed,currenttime, NULL);
  }

  helicsFederateFinalize(vfed,NULL);
  printf("PI SENDER: Federate finalized\n");

  helicsFederateFree(vfed);
  while(helicsBrokerIsConnected(broker)) {
#ifdef _MSC_VER
      Sleep(100);
#else
    usleep(100000); /* Sleep for 100 millisecond */
#endif
  }
  printf("PI SENDER: Broker disconnected\n");
  helicsBrokerFree(broker);
  helicsCloseLibrary();
  return(0);
}

